lib/APPTurtle/src/Transports/CustomRosTransportTcp.cpp

```diff
@@ -0,0 +1,353 @@
+
+/******************************************************************************
+ * Prototypes
+ *****************************************************************************/
+
+/******************************************************************************
+ * Local Variables
+ *****************************************************************************/
+
+const String CustomRosTransportTcp::m_protocolName("TCP");
+
+const CustomRosTransportTcp::ReadFunc
+CustomRosTransportTcp::m_readFunction[InputState::INPUT_STATE_MAX] = 
+{
+    &CustomRosTransportTcp::readSizePrefix, 
+    &CustomRosTransportTcp::readPendingSizePrefix,
+    &CustomRosTransportTcp::readPayload, 
+    &CustomRosTransportTcp::readFinish
+};
+
+/******************************************************************************
+ * Public Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Protected Methods
+ *****************************************************************************/
+
+/******************************************************************************
+ * Private Methods
+ *****************************************************************************/
+
+bool CustomRosTransportTcp::open(void)
+{
+    bool isOpen = false;
+
+    if (0U == m_tcpClient.connect(m_address, m_port))
+    {
+        LOG_ERROR("TCP connect error");
+    }
+    else
+    {
+        isOpen = true;
+    }
+
+    return isOpen;
+}
+
+bool CustomRosTransportTcp::close()
+{
+    m_tcpClient.stop();
+    return true;
+}
+
+size_t CustomRosTransportTcp::write(const uint8_t* buffer, size_t size, uint8_t* errorCode)
+{
+    size_t sent = 0U;
+
+    if ((nullptr == buffer) || (0U == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        uint8_t lengthPrefix[2];
+        lengthPrefix[0] = static_cast<uint8_t>(size & 0xFFU);
+        lengthPrefix[1] = static_cast<uint8_t>((size >> 8U) & 0xFFU);
+
+        if (sizeof(lengthPrefix) != m_tcpClient.write(lengthPrefix, sizeof(lengthPrefix)))
+        {
+            LOG_ERROR("Write prefix error (size=%zu, sent=%zu)", size, sent);
+            *errorCode = 1U;
+        }
+        else if ((sent = m_tcpClient.write(buffer, size)) != size)
+        {
+            LOG_ERROR("Write data error (size=%zu, sent=%zu)", size, sent);
+            *errorCode = 1U;
+        }
+        else
+        {
+            *errorCode = 0U;
+        }
+    }
+
+    return sent;
+}
+
+size_t CustomRosTransportTcp::read(uint8_t* buffer, size_t size, int timeout, uint8_t* errorCode)
+{
+    size_t readBytes = 0U;
+
+    if ((nullptr == buffer) || (0U == size) || (nullptr == errorCode))
+    {
+        LOG_ERROR("One or more parameters are invalid.");
+    }
+    else
+    {
+        bool loop = true;
+
+        /**
```

Comment:
Doxygen commment inside function body will destroy the generated documentation.

Reply:
Ups